packages/ra-core/src/reducer/admin/resource/data.js

```diff
@@ -60,40 +57,43 @@ const addRecords = addRecordsFactory(getFetchedAt);
 const initialState = {};
 Object.defineProperty(initialState, 'fetchedAt', { value: {} }); // non enumerable by default
 
-export default (previousState = initialState, { type, payload, meta }) => {
-    if (type === CRUD_UPDATE_OPTIMISTIC) {
-        const updatedRecord = { ...previousState[payload.id], ...payload.data };
-        return addRecords([updatedRecord], previousState);
-    }
-    if (type === CRUD_UPDATE_MANY_OPTIMISTIC) {
-        const updatedRecords = payload.ids
-            .reduce((records, id) => records.concat(previousState[id]), [])
-            .map(record => ({ ...record, ...payload.data }));
-        return addRecords(updatedRecords, previousState);
-    }
-    if (type === CRUD_DELETE_OPTIMISTIC) {
-        const { [payload.id]: removed, ...newState } = previousState;
+export default (previousState = initialState, { payload, meta }) => {
+    if (meta && meta.optimistic) {
+        if (meta.fetch === UPDATE) {
+            const updatedRecord = { ...previousState[payload.id], ...payload.data };
+            return addRecords([updatedRecord], previousState);
+        }
+        if (meta.fetch === UPDATE_MANY) {
+            const updatedRecords = payload.ids
+                .reduce((records, id) => records.concat(previousState[id]), [])
```

Comment:
Unless previousState[id] returns an array you can use a map here. And in this case you can do it in a single map

```js
.map(id => ({
    ...previousState[id],
    ...payload.data,
}));
```

Reply:
I think the `reduce` has another purpose: avoid updating non-existing ids.

Reply:
Even so, it makes no difference. We create a new array any way from the ids in the payload.
And we will always merge the previousState with the data from the payloads, so we might as well do it in a single map.

Reply:
agreed